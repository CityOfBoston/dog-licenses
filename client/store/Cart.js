// @flow

import { observable, computed, action } from 'mobx';

import type { DeathCertificate } from '../types';

export class CartItem {
  @observable.ref cert: DeathCertificate;
  @observable quantity: number;
}

export default class Cart {
  @observable items = [];

  @computed get size(): number {
    return this.items.reduce((acc, item) => acc + item.quantity, 0);
  }

  @action add(cert: DeathCertificate, quantity: number) {
    const existingItem = this.items.find((item) => item.cert.id === cert.id);
    if (existingItem) {
      existingItem.quantity += quantity;
    } else {
      const item = new CartItem();
      item.cert = cert;
      item.quantity = quantity;

      this.items.push(item);
    }
  }

  @action setQuantity(certId: string, quantity: number) {
    const existingItem = this.items.find((item) => item.cert.id === certId);
    if (existingItem) {
      existingItem.quantity = quantity;
    }
  }

  @action remove(certId: string) {
    const idx = this.items.findIndex(({ cert }) => cert.id === certId);
    if (idx !== -1) {
      this.items.splice(idx, 1);
    }
  }
}
