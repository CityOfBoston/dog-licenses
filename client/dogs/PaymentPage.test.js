// @flow
import React from 'react';
import { runInAction } from 'mobx';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import Cart from '../store/Cart';
import {
  TYPICAL_LICENSE,
  PENDING_LICENSE,
} from '../../fixtures/client/dog-licenses';

import PaymentPage from './PaymentPage';

describe('rendering', () => {
  let cart;

  beforeEach(() => {
    cart = new Cart();
  });

  it('renders loading for unhydrated cart', () => {
    runInAction(() => {
      cart.pendingFetches = 2;
      cart.items = ([
        {
          id: TYPICAL_LICENSE.id,
          cert: null,
          quantity: 5,
        },
        {
          id: PENDING_LICENSE.id,
          cert: null,
          quantity: 3,
        },
      ]: any);
    });

    expect(
      renderer.create(<PaymentPage cart={cart} />).toJSON(),
    ).toMatchSnapshot();
  });

  it('renders a hydrated cart', () => {
    cart.add(TYPICAL_LICENSE, 5);
    cart.add(PENDING_LICENSE, 2);

    expect(
      renderer.create(<PaymentPage cart={cart} />).toJSON(),
    ).toMatchSnapshot();
  });
});

describe('submit', () => {
  let cart;
  let wrapper;

  beforeEach(() => {
    cart = new Cart();
    cart.add(TYPICAL_LICENSE, 5);
    cart.add(PENDING_LICENSE, 2);

    wrapper = shallow(<PaymentPage cart={cart} />);
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it('submits', () => {
    const form = wrapper.find('form');
    const preventDefault = jest.fn();

    form.simulate('submit', { preventDefault });
    expect(preventDefault).toHaveBeenCalled();
  });
});
