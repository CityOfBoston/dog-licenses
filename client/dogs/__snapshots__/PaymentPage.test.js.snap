// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders a hydrated cart 1`] = `
<div>
  <nav
    className="nv-s nv-s--sticky"
    data-jsx={792433154}
  >
    <div
      className="nv-s-l bar"
      data-jsx={792433154}
    >
      <a
        className="nv-s-l-b back-link"
        data-jsx={792433154}
        href="/dogs"
        onClick={[Function]}
      >
        Back to Lookup
      </a>
      <a
        className="cart-link"
        data-jsx={792433154}
        href="/dogs/checkout"
        onClick={[Function]}
      >
        7
      </a>
    </div>
  </nav>
  <div
    className="p-a300"
    data-jsx={1303275230}
  >
    <div
      className="sh sh--b0 m-v300"
      data-jsx={1303275230}
    >
      <h1
        className="sh-title"
        data-jsx={1303275230}
      >
        Pay and Finish
      </h1>
    </div>
    <button
      className="sel-c sel-c--fw"
      data-jsx={1303275230}
      type="button"
    >
      <div
        className="sel-f"
        data-jsx={1303275230}
      >
        <div
          className="t--sans tt-u lnk"
          data-jsx={1303275230}
        >
          Order Details
        </div>
        <div
          className="t--info"
          data-jsx={1303275230}
        >
          7
           
          licenses
           Ã— $
          14
           + fee = $
          100.70
        </div>
      </div>
    </button>
  </div>
  <form
    acceptCharset="UTF-8"
    action="/dogs/payment"
    className="m-v300"
    data-jsx={3785310163}
    method="post"
    onSubmit={[Function]}
  >
    <div
      className="b--w p-a500 field-container"
      data-jsx={3785310163}
    >
      <fieldset
        data-jsx={3785310163}
      >
        <legend
          className="stp m-b300"
          data-jsx={3785310163}
        >
          Contact
        </legend>
        <div
          className="txt"
          data-jsx={3785310163}
        >
          <label
            className="txt-l txt-l--mt000"
            data-jsx={3785310163}
            htmlFor="contact-email"
          >
            Email Address
          </label>
          <input
            className="txt-f txt-f--100"
            data-jsx={3785310163}
            id="contact-email"
            name="email"
            placeholder="Email address"
            type="email"
          />
        </div>
        <div
          className="txt"
          data-jsx={3785310163}
        >
          <label
            className="txt-l"
            data-jsx={3785310163}
            htmlFor="contact-phone"
          >
            Phone Number
          </label>
          <input
            className="txt-f txt-f--100"
            data-jsx={3785310163}
            id="contact-phone"
            name="phone"
            placeholder="Phone number"
            type="phone"
          />
        </div>
      </fieldset>
    </div>
    <div
      className="b--w p-a500 field-container"
      data-jsx={3785310163}
    >
      <fieldset
        data-jsx={3785310163}
      >
        <legend
          className="stp m-b300"
          data-jsx={3785310163}
        >
          Credit Card Information
        </legend>
        <div
          className="txt"
          data-jsx={3785310163}
        >
          <label
            className="txt-l txt-l--mt000"
            data-jsx={3785310163}
            htmlFor="card-name"
          >
            Name on Card
          </label>
          <input
            className="txt-f"
            data-jsx={3785310163}
            id="card-name"
            name="card-name"
            placeholder="Name on card"
            type="text"
          />
        </div>
        <div
          className="txt"
          data-jsx={3785310163}
        >
          <label
            className="txt-l"
            data-jsx={3785310163}
            htmlFor="card-number"
          >
            Credit Card Number
          </label>
          <input
            className="txt-f"
            data-jsx={3785310163}
            id="card-number"
            name="card-number"
            placeholder="Credit Card Number"
            type="text"
          />
        </div>
        <div
          className="txt"
          data-jsx={3785310163}
        >
          <label
            className="txt-l"
            data-jsx={3785310163}
            htmlFor="card-date"
          >
            Expiration Date
          </label>
          <input
            className="txt-f txt-f--50"
            data-jsx={3785310163}
            id="card-date"
            name="card-date"
            placeholder="mm/yy"
          />
        </div>
      </fieldset>
    </div>
    <div
      className="b--w p-a500 field-container"
      data-jsx={3785310163}
    >
      <fieldset
        data-jsx={3785310163}
      >
        <legend
          className="stp m-b300"
          data-jsx={3785310163}
        >
          Billing Information
        </legend>
        <div
          className="txt"
          data-jsx={3785310163}
        >
          <label
            className="txt-l txt-l--mt000"
            data-jsx={3785310163}
            htmlFor="billing-address-1"
          >
            Address Line 1
          </label>
          <input
            className="txt-f"
            data-jsx={3785310163}
            id="billing-address-1"
            name="billing-address-1"
            placeholder="Address Line 1"
            type="text"
          />
        </div>
        <div
          className="txt"
          data-jsx={3785310163}
        >
          <label
            className="txt-l txt-l--mt000"
            data-jsx={3785310163}
            htmlFor="billing-address-2"
          >
            Address Line 2
          </label>
          <input
            className="txt-f"
            data-jsx={3785310163}
            id="billing-address-2"
            name="billing-address-2"
            placeholder="Address Line 2"
            type="text"
          />
        </div>
        <div
          className="txt"
          data-jsx={3785310163}
        >
          <label
            className="txt-l"
            data-jsx={3785310163}
            htmlFor="billing-city"
          >
            City
          </label>
          <input
            className="txt-f"
            data-jsx={3785310163}
            id="billing-city"
            name="billing-city"
            placeholder="City"
            type="text"
          />
        </div>
        <div
          className="txt"
          data-jsx={3785310163}
        >
          <label
            className="txt-l"
            data-jsx={3785310163}
            htmlFor="billing-state"
          >
            State
          </label>
          <input
            className="txt-f txt-f--50"
            data-jsx={3785310163}
            id="billing-state"
            name="billing-state"
            placeholder="State"
          />
        </div>
        <div
          className="txt"
          data-jsx={3785310163}
        >
          <label
            className="txt-l"
            data-jsx={3785310163}
            htmlFor="billing-zip"
          >
            ZIP Code
          </label>
          <input
            className="txt-f txt-f--50"
            data-jsx={3785310163}
            id="billing-zip"
            name="billing-zip"
            placeholder="ZIP code"
          />
        </div>
      </fieldset>
    </div>
    <div
      className="b--w p-a500 field-container"
      data-jsx={3785310163}
    >
      <fieldset
        data-jsx={3785310163}
      >
        <legend
          className="stp m-b300"
          data-jsx={3785310163}
        >
          Shipping Information
        </legend>
        <div
          className="txt"
          data-jsx={3785310163}
        >
          <label
            className="txt-l txt-l--mt000"
            data-jsx={3785310163}
            htmlFor="shipping-address-1"
          >
            Address Line 1
          </label>
          <input
            className="txt-f"
            data-jsx={3785310163}
            id="shipping-address-1"
            name="shipping-address-1"
            placeholder="Address Line 1"
            type="text"
          />
        </div>
        <div
          className="txt"
          data-jsx={3785310163}
        >
          <label
            className="txt-l txt-l--mt000"
            data-jsx={3785310163}
            htmlFor="shipping-address-2"
          >
            Address Line 2
          </label>
          <input
            className="txt-f"
            data-jsx={3785310163}
            id="shipping-address-2"
            name="shipping-address-2"
            placeholder="Address Line 2"
            type="text"
          />
        </div>
        <div
          className="txt"
          data-jsx={3785310163}
        >
          <label
            className="txt-l"
            data-jsx={3785310163}
            htmlFor="shipping-city"
          >
            City
          </label>
          <input
            className="txt-f"
            data-jsx={3785310163}
            id="shipping-city"
            name="shipping-city"
            placeholder="City"
            type="text"
          />
        </div>
        <div
          className="txt"
          data-jsx={3785310163}
        >
          <label
            className="txt-l"
            data-jsx={3785310163}
            htmlFor="shipping-state"
          >
            State
          </label>
          <input
            className="txt-f txt-f--50"
            data-jsx={3785310163}
            id="shipping-state"
            name="shipping-state"
            placeholder="State"
          />
        </div>
        <div
          className="txt"
          data-jsx={3785310163}
        >
          <label
            className="txt-l"
            data-jsx={3785310163}
            htmlFor="shipping-zip"
          >
            ZIP Code
          </label>
          <input
            className="txt-f txt-f--50"
            data-jsx={3785310163}
            id="shipping-zip"
            name="shipping-zip"
            placeholder="ZIP code"
          />
        </div>
      </fieldset>
    </div>
    <div
      className="g p-a500 b--w"
      data-jsx={3785310163}
    >
      <button
        className="g--3 btn m-v500"
        data-jsx={3785310163}
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;

exports[`rendering renders loading for unhydrated cart 1`] = `
<div>
  <nav
    className="nv-s nv-s--sticky"
    data-jsx={792433154}
  >
    <div
      className="nv-s-l bar"
      data-jsx={792433154}
    >
      <a
        className="nv-s-l-b back-link"
        data-jsx={792433154}
        href="/dogs"
        onClick={[Function]}
      >
        Back to Lookup
      </a>
      <a
        className="cart-link"
        data-jsx={792433154}
        href="/dogs/checkout"
        onClick={[Function]}
      >
        8
      </a>
    </div>
  </nav>
  <div
    className="p-a300"
    data-jsx={1303275230}
  >
    <div
      className="sh sh--b0 m-v300"
      data-jsx={1303275230}
    >
      <h1
        className="sh-title"
        data-jsx={1303275230}
      >
        Pay and Finish
      </h1>
    </div>
    <button
      className="sel-c sel-c--fw"
      data-jsx={1303275230}
      type="button"
    >
      <div
        className="sel-f"
        data-jsx={1303275230}
      >
        <div
          className="t--sans tt-u lnk"
          data-jsx={1303275230}
        >
          Order Details
        </div>
        <div
          className="t--info"
          data-jsx={1303275230}
        >
          8
           
          licenses
           Ã— $
          14
           + fee = $
          115.09
        </div>
      </div>
    </button>
  </div>
  <form
    acceptCharset="UTF-8"
    action="/dogs/payment"
    className="m-v300"
    data-jsx={3785310163}
    method="post"
    onSubmit={[Function]}
  >
    <div
      className="b--w p-a500 field-container"
      data-jsx={3785310163}
    >
      <fieldset
        data-jsx={3785310163}
      >
        <legend
          className="stp m-b300"
          data-jsx={3785310163}
        >
          Contact
        </legend>
        <div
          className="txt"
          data-jsx={3785310163}
        >
          <label
            className="txt-l txt-l--mt000"
            data-jsx={3785310163}
            htmlFor="contact-email"
          >
            Email Address
          </label>
          <input
            className="txt-f txt-f--100"
            data-jsx={3785310163}
            id="contact-email"
            name="email"
            placeholder="Email address"
            type="email"
          />
        </div>
        <div
          className="txt"
          data-jsx={3785310163}
        >
          <label
            className="txt-l"
            data-jsx={3785310163}
            htmlFor="contact-phone"
          >
            Phone Number
          </label>
          <input
            className="txt-f txt-f--100"
            data-jsx={3785310163}
            id="contact-phone"
            name="phone"
            placeholder="Phone number"
            type="phone"
          />
        </div>
      </fieldset>
    </div>
    <div
      className="b--w p-a500 field-container"
      data-jsx={3785310163}
    >
      <fieldset
        data-jsx={3785310163}
      >
        <legend
          className="stp m-b300"
          data-jsx={3785310163}
        >
          Credit Card Information
        </legend>
        <div
          className="txt"
          data-jsx={3785310163}
        >
          <label
            className="txt-l txt-l--mt000"
            data-jsx={3785310163}
            htmlFor="card-name"
          >
            Name on Card
          </label>
          <input
            className="txt-f"
            data-jsx={3785310163}
            id="card-name"
            name="card-name"
            placeholder="Name on card"
            type="text"
          />
        </div>
        <div
          className="txt"
          data-jsx={3785310163}
        >
          <label
            className="txt-l"
            data-jsx={3785310163}
            htmlFor="card-number"
          >
            Credit Card Number
          </label>
          <input
            className="txt-f"
            data-jsx={3785310163}
            id="card-number"
            name="card-number"
            placeholder="Credit Card Number"
            type="text"
          />
        </div>
        <div
          className="txt"
          data-jsx={3785310163}
        >
          <label
            className="txt-l"
            data-jsx={3785310163}
            htmlFor="card-date"
          >
            Expiration Date
          </label>
          <input
            className="txt-f txt-f--50"
            data-jsx={3785310163}
            id="card-date"
            name="card-date"
            placeholder="mm/yy"
          />
        </div>
      </fieldset>
    </div>
    <div
      className="b--w p-a500 field-container"
      data-jsx={3785310163}
    >
      <fieldset
        data-jsx={3785310163}
      >
        <legend
          className="stp m-b300"
          data-jsx={3785310163}
        >
          Billing Information
        </legend>
        <div
          className="txt"
          data-jsx={3785310163}
        >
          <label
            className="txt-l txt-l--mt000"
            data-jsx={3785310163}
            htmlFor="billing-address-1"
          >
            Address Line 1
          </label>
          <input
            className="txt-f"
            data-jsx={3785310163}
            id="billing-address-1"
            name="billing-address-1"
            placeholder="Address Line 1"
            type="text"
          />
        </div>
        <div
          className="txt"
          data-jsx={3785310163}
        >
          <label
            className="txt-l txt-l--mt000"
            data-jsx={3785310163}
            htmlFor="billing-address-2"
          >
            Address Line 2
          </label>
          <input
            className="txt-f"
            data-jsx={3785310163}
            id="billing-address-2"
            name="billing-address-2"
            placeholder="Address Line 2"
            type="text"
          />
        </div>
        <div
          className="txt"
          data-jsx={3785310163}
        >
          <label
            className="txt-l"
            data-jsx={3785310163}
            htmlFor="billing-city"
          >
            City
          </label>
          <input
            className="txt-f"
            data-jsx={3785310163}
            id="billing-city"
            name="billing-city"
            placeholder="City"
            type="text"
          />
        </div>
        <div
          className="txt"
          data-jsx={3785310163}
        >
          <label
            className="txt-l"
            data-jsx={3785310163}
            htmlFor="billing-state"
          >
            State
          </label>
          <input
            className="txt-f txt-f--50"
            data-jsx={3785310163}
            id="billing-state"
            name="billing-state"
            placeholder="State"
          />
        </div>
        <div
          className="txt"
          data-jsx={3785310163}
        >
          <label
            className="txt-l"
            data-jsx={3785310163}
            htmlFor="billing-zip"
          >
            ZIP Code
          </label>
          <input
            className="txt-f txt-f--50"
            data-jsx={3785310163}
            id="billing-zip"
            name="billing-zip"
            placeholder="ZIP code"
          />
        </div>
      </fieldset>
    </div>
    <div
      className="b--w p-a500 field-container"
      data-jsx={3785310163}
    >
      <fieldset
        data-jsx={3785310163}
      >
        <legend
          className="stp m-b300"
          data-jsx={3785310163}
        >
          Shipping Information
        </legend>
        <div
          className="txt"
          data-jsx={3785310163}
        >
          <label
            className="txt-l txt-l--mt000"
            data-jsx={3785310163}
            htmlFor="shipping-address-1"
          >
            Address Line 1
          </label>
          <input
            className="txt-f"
            data-jsx={3785310163}
            id="shipping-address-1"
            name="shipping-address-1"
            placeholder="Address Line 1"
            type="text"
          />
        </div>
        <div
          className="txt"
          data-jsx={3785310163}
        >
          <label
            className="txt-l txt-l--mt000"
            data-jsx={3785310163}
            htmlFor="shipping-address-2"
          >
            Address Line 2
          </label>
          <input
            className="txt-f"
            data-jsx={3785310163}
            id="shipping-address-2"
            name="shipping-address-2"
            placeholder="Address Line 2"
            type="text"
          />
        </div>
        <div
          className="txt"
          data-jsx={3785310163}
        >
          <label
            className="txt-l"
            data-jsx={3785310163}
            htmlFor="shipping-city"
          >
            City
          </label>
          <input
            className="txt-f"
            data-jsx={3785310163}
            id="shipping-city"
            name="shipping-city"
            placeholder="City"
            type="text"
          />
        </div>
        <div
          className="txt"
          data-jsx={3785310163}
        >
          <label
            className="txt-l"
            data-jsx={3785310163}
            htmlFor="shipping-state"
          >
            State
          </label>
          <input
            className="txt-f txt-f--50"
            data-jsx={3785310163}
            id="shipping-state"
            name="shipping-state"
            placeholder="State"
          />
        </div>
        <div
          className="txt"
          data-jsx={3785310163}
        >
          <label
            className="txt-l"
            data-jsx={3785310163}
            htmlFor="shipping-zip"
          >
            ZIP Code
          </label>
          <input
            className="txt-f txt-f--50"
            data-jsx={3785310163}
            id="shipping-zip"
            name="shipping-zip"
            placeholder="ZIP code"
          />
        </div>
      </fieldset>
    </div>
    <div
      className="g p-a500 b--w"
      data-jsx={3785310163}
    >
      <button
        className="g--3 btn m-v500"
        data-jsx={3785310163}
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;
